Classes/common/Encryption/Keychain/CDTEncryptionKeychainStorage.m

```diff
@@ -0,0 +1,198 @@
+    NSMutableDictionary *lookupDict =
+        [CDTEncryptionKeychainStorage genericPwLookupDictWithService:self.service
+                                                             account:self.account];
+
+    OSStatus err = SecItemCopyMatching((__bridge CFDictionaryRef)lookupDict, (void *)&data);
+    if (err == noErr) {
+        result = ((data != nil) && (data.length > 0));
+
+        if (!result) {
+            CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"Found a match in keychain, but it was empty");
+        }
+    } else if (err == errSecItemNotFound) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"DPK doc not found in keychain");
+    } else {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Error getting DPK doc from keychain, SecItemCopyMatching returned: %d", err);
+    }
+
+    return result;
+}
+
+#pragma mark - Private class methods
++ (NSMutableDictionary *)genericPwLookupDictWithService:(NSString *)service
+                                                account:(NSString *)account
+{
+    NSMutableDictionary *genericPasswordQuery = [[NSMutableDictionary alloc] init];
+    
+    [genericPasswordQuery setObject:(__bridge id)kSecClassGenericPassword
+                             forKey:(__bridge id)kSecClass];
+    [genericPasswordQuery setObject:service forKey:(__bridge id<NSCopying>)(kSecAttrService)];
+    [genericPasswordQuery setObject:account forKey:(__bridge id<NSCopying>)(kSecAttrAccount)];
+
+    // Use the proper search constants, return only the attributes of the first match.
+    [genericPasswordQuery setObject:(__bridge id)kSecMatchLimitOne
+                             forKey:(__bridge id<NSCopying>)(kSecMatchLimit)];
+    [genericPasswordQuery setObject:(__bridge id)kCFBooleanFalse
+                             forKey:(__bridge id<NSCopying>)(kSecReturnAttributes)];
+    [genericPasswordQuery setObject:(__bridge id)kCFBooleanTrue
+                             forKey:(__bridge id<NSCopying>)(kSecReturnData)];
+    
+    return genericPasswordQuery;
+}
+
++ (NSMutableDictionary *)genericPwStoreDictWithService:(NSString *)service
+                                               account:(NSString *)account
+                                                  data:(NSData *)data
+{
+    NSMutableDictionary *genericPasswordQuery = [[NSMutableDictionary alloc] init];
+    
+    [genericPasswordQuery setObject:(__bridge id)kSecClassGenericPassword
+                             forKey:(__bridge id)kSecClass];
+    [genericPasswordQuery setObject:service forKey:(__bridge id<NSCopying>)(kSecAttrService)];
+    [genericPasswordQuery setObject:account forKey:(__bridge id<NSCopying>)(kSecAttrAccount)];
+
+    [genericPasswordQuery setObject:data forKey:(__bridge id<NSCopying>)(kSecValueData)];
+
+#warning This does not look like the best option
+    [genericPasswordQuery setObject:(__bridge id)(kSecAttrAccessibleAlways)
+                             forKey:(__bridge id<NSCopying>)(kSecAttrAccessible)];
```

Comment:
We have a bigger problem here, the code I wrote so far is valid only for iOS because the Mac OS Keychain Services API uses a different paradigm (check this [documentation](https://developer.apple.com/library/mac/documentation/Security/Conceptual/keychainServConcepts/02concepts/concepts.html), under title *Keychain Services Ease of Use*).

This code has to be rewritten for Mac OS X.